### ISSUE TITLE
False-positive `experiment_not_enabled` for null safety with Dart 2.9.1

### ISSUE THREAD

**Issue by craiglabenz** [flutter/flutter#64051]
## Steps to Reproduce

1. Run `flutter create <name>`.
2. Change the `_incrementCounter` signature to be `_incrementCounter(int? val)`, and see the red squiggle underneath `?`. At this point, this is expected because as of Flutter 1.20.2, the generated `pubpsec.yaml` specifies a minimum SDK version of 2.7.
3. Update `pubspec.yaml` to specify a minimum SDK version of 2.9, as per the helpful warning.
4. The red squiggle does not go away.

**Expected results:** Null-aware syntax should work when the SDK version is specified to be >= 2.9.0.

**Actual results:** Null-aware syntax was still warned against.

## Flutter Doctor

```
$ flutter doctor -v
[✓] Flutter (Channel stable, 1.20.2, on Mac OS X 10.15.6 19G73, locale en-US)
    • Flutter version 1.20.2 at /Users/craiglabenz/flutter
    • Framework revision bbfbf1770c (5 days ago), 2020-08-13 08:33:09 -0700
    • Engine revision 9d5b21729f
    • Dart version 2.9.1

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/craiglabenz/Library/Android/sdk
    • Platform android-29, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.8.4

[!] Android Studio (version 3.4)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[✓] VS Code (version 1.48.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.13.2

 
[✓] Connected device (1 available)            
    • iPhone 11 Pro (mobile) • 1A9A1101-CCEA-4D89-8AC4-3E321A40369A • ios •
      com.apple.CoreSimulator.SimRuntime.iOS-13-6 (simulator)

! Doctor found issues in 1 category.
```

</details>

**Comment by filiph** [comment]
I think that, even after you upgrade to 2.9, you still have to enable null safety. And it's probably not going to work with Flutter just yet. Is something forcing you to use null safety or is this you trying it out?

**Comment by craiglabenz** [comment]
Thanks for the response, @filiph!

I am not aware of having taken any specific steps to opt-in to null-safety. In the above example, I was able to reproduce on a fresh project with just that minimal change, and didn't make any edits to the project other than what's detailed.

Are there possibly system-wide flags that could influence the situation?

**Comment by filiph** [comment]
Sorry, I still don't understand: are you trying to use null safety, or is there something forcing you to use it?

If the latter, that's a bug. If the former, you'll have to jump through hoops (possibly walls) to get it working in Flutter. See the "Try it now" section of [this announcement](https://medium.com/dartlang/announcing-sound-null-safety-defd2216a6f3) for how to enable the null safety experiment (but that only covers Dart command line programs, not Flutter).

**Comment by craiglabenz** [comment]
Sorry for the delay, but I'm ready to clarify. There's a few layers here, [so bear with me](https://contenthub-static.grammarly.com/blog/wp-content/uploads/2017/02/BEAR-WITH-ME-.jpg).

The change you referenced:

> 2. Change the _incrementCounter signature to be _incrementCounter(int? val)

was just me writing the bug recreate steps. So, the next obvious question is why I was recreating that, given that I don't want to jump through walls, or even just hoops, to tackle non-nullability in Flutter ahead of everyone else.

The answer there is that I unexpectedly received this runtime error while registering a listener with a stream of ints:

```
Unhandled error type '(int) => void' is not a subtype of '((int) => void)?'
```

Oddly, when I jumped-to-source my `.listen()` call, it took me to this path on my local machine:

```
~/flutter/bin/cache/pkg/sky_engine/lib/async/stream.dart
```

That's surprising, because I definitely expected to jump to:

```
~/flutter/bin/cache/dart-sdk/lib/async/stream.dart
```

---

HOWEVER, interesting as all that is, it may be immaterial, because both versions of `Stream` have the same null-safety-aware signature for `listen`:

```dart
StreamSubscription<T> listen(void onData(T event)?,
    {Function? onError, void onDone()?, bool? cancelOnError});
```

And thus I concluded that my local Flutter environment (Flutter 1.20.2 and Dart 1.9.1) was bumping into the world of null-safety ahead of schedule.

Please let me know if anything about this novella is still confusing!

**Comment by pedromassangocode** [comment]
Hi @craiglabenz 
Can you please provide a minimal reproducible code sample that caused this:

> The answer there is that I unexpectedly received this runtime error while registering a listener with a stream of ints:
> 
> ```
> Unhandled error type '(int) => void' is not a subtype of '((int) => void)?'
> ```

**Comment by craiglabenz** [comment]
Hey @pedromassango! I am still investigating how to recreate that exact error with minimal changes off the starter app. So far, [I made this sample](https://gist.github.com/craiglabenz/d0b67a3a59db8213d7181b7954fa0193), which as far as I can tell is equivalent to the error I'm seeing in my real project, but it passes.

This must mean there's something funky with my project or environment 🤔

**Comment by craiglabenz** [comment]
OKAY. As we were beginning to conclude, I'm pretty sure this is all my doing.

My actual situation was not just `int`s, but a list of generic objects, themselves with a type. My full, actual error, (these relevant details I only just noticed), is:

```
Unhandled error type: '(Message) => void' is not a subtype of '((BaseMessage<String>) => void)?'
```

Sooooo I'm ~95% sure the error has to do with insufficiently defining my type on `Message` (`<String>`), and that the trailing `?` never had anything to do with it.

Closing for now unless the above theory ends up being false.
